### PR TITLE
Update instance segmentation models

### PR DESCRIPTION
**Pull request by DmitriySidnev**
Added new instance segmentation models based on `mmdetection`.
@snosov1, @Ilya-Krylov, @druzhkov-paul

### CODE REVIEW

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+_base_ = '../../../external/mmdetection/configs/mask_rcnn/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco.py'
```
Comment (druzhkov-paul) [inline pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: Is it Ok to use such relative paths?

Comment (druzhkov-paul) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: Shouldn't these models be compliant with Model Templates? Do they?

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0091.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+data = dict(
+    train=dict(
+        ann_file=data_root + 'annotations/instances_train2017.json',
+        img_prefix=data_root + 'train2017/'),
+    val=dict(
+        ann_file=data_root + 'annotations/instances_val2017.json',
+        img_prefix=data_root + 'val2017/'),
+    test=dict(
+        ann_file=data_root + 'annotations/instances_val2017.json',
+        img_prefix=data_root + 'val2017/'))
```
Comment (druzhkov-paul) [inline pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0091.py:12]: Is it just the public model from mmdetection?

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        feat_channels=128),
+    roi_head=dict(
+        bbox_roi_extractor=dict(
+            out_channels=128),
+        bbox_head=dict(
+            in_channels=128,
+            fc_out_channels=512,
+            roi_feat_size=7),
+        mask_roi_extractor=dict(
+            roi_layer=dict(type='RoIAlign', out_size=7, sample_num=0),
```
Comment (druzhkov-paul) [inline pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py:19]: 1. What's the quality degradation at the OpenVINO side?
1. How does mmdetection RoIAlign compare with torchvision one at the PyTorch & OpenVINO sides?
1. Is it the aligned version?

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py (hunk 1 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: These models have to be complaint with Model Templates (https://github.com/openvinotoolkit/training_extensions/pull/335/).

During runtime of a platform model folders are usually copied outside of OTE repo folder, so that such relative paths are forbidden.

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev sorry for not mentioning this earlier but current text-spotting depends on segmentoly. So we can remove it as soon as I replace current version. Please turn it back.

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0091.py (hunk 2 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0091.py:12]: Yes.

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py (hunk 3 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py:19]: 1. Metrics are the same in `Pytorch` and `OpenVINO` for first three models (at least in v.2020.4) when all images are resized to a target resolution without keeping ratio of width and height. For the two last models there is a quality degradation because of image preprocessing where is set to True `keep_ratio` option and some images are stretched by height when OpenVINO model expects horizontal images. I forgot about it.
  2. I do not know, did not try to change the type. It just works without any issues. 
  3. What do you mean?

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py (hunk 1 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: I thought that we should use submodule placed in `training_extensions/external/mmdetection`. If it is true, then relative paths of base configuration files are correct, I have checked training and validation commands. Should it be full config-files without `_base_` option? In config may be incorrect relative path of `data_root` directory which depends on a running directory.

Comment (Ilya-Krylov) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: There are two cases at least we have to support both.
1. Training from repo. In that case what you have now is OK.
2. Training in a platform where model folder (config.py and other stuff) can be moved to some place and mmdetection can be  pre-built in some docker image. Here we can't use `_base_`. `data_root` in this case will be definitely changed. So please copy-paste content of `_base_` config right here.

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev Each model must have corresponding 
* `template.yaml` file 
* `train.py`, `eval.py`, `export.py`, `quantize.py` with common interface defined in https://github.com/openvinotoolkit/training_extensions/pull/335

you can use MMDETECTION_DIR environment variable to point to mmdetection

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py (hunk 3 above)
Comment (druzhkov-paul) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0228.py:19]: 1. Cool. It's not what I would expect though.

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py (hunk 1 above)
Comment (druzhkov-paul) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py:1]: I guess you can keep base config in a separate file, but have it here. @Ilya-Krylov am I right?

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, if `segmentoly` will be back, should I use the same directory for both set of models or move new models to another directory?

File: pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py (hunk 1 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation/configs/instance-segmentation-0002.py]: fixed with complete configuration files.

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev I think it is ok to use the same dir, I hope I'll replace text spotting soon.

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, interfaces for  `train.py`, `eval.py`, `export.py`, `quantize.py` are described [here](https://github.com/openvinotoolkit/training_extensions/tree/model_template/pytorch_toolkit/ote)?

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, I have updated scripts and tests for instance segmentation task. Could you please take a look one more time?

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+arg_parser: DefaultArgParser
+arg_converter: InstanceSegmentationArgsConverter
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml:2]: Having mmdetection counterparts is not enough?

File: pytorch_toolkit/instance_segmentation_2/tests/run_export_tests.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Copyright (C) 2020 Intel Corporation
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/instance_segmentation_2/tests/run_export_tests.py:1]: this file is not needed

File: pytorch_toolkit/instance_segmentation_2/tests/run_inference_tests.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# Copyright (C) 2020 Intel Corporation
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/instance_segmentation_2/tests/run_inference_tests.py:1]: this file is not needed

File: pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            if snapshot.split('.')[-1] == 'bin':
+                snapshot = '.'.join(snapshot.split('.')[:-1]) + '.xml'
+            tool = 'test_exported.py'
+        else:
+            tool = 'test.py'
+
+        subprocess.run(
+            f'python {MMDETECTION_TOOLS}/{tool}'
+            f' {config_path} {snapshot}'
+            f' --out {res_pkl} --eval bbox segm'
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py:73]: I guess metrics types can be obtained in different ways, one of them is to load it from configuration file

File: pytorch_toolkit/ote/ote/modules/arg_converters/instance_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+"""
+
+from .base import BaseArgConverter
+from ..registry import ARG_CONVERTERS
+
+
+@ARG_CONVERTERS.register_module()
+class InstanceSegmentationArgsConverter(BaseArgConverter):
+    train_update_args_map = {
+        'train_ann_files': 'data.train.ann_file',
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/instance_segmentation.py:24]: this is also can be taken from mmdetection args converter. You can modify configuration files by modifying data.train section adding Repeat dataset with value=1. It will not change your config.

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev if there are only two differences between mmdetection that are listed above, I think we should use the same classes and avoid code duplication. Moreover detection and segmentation models come from the same framework.

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml (hunk 4 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml:2]: `ArgsConverter` for instance segmentation has a couple of differences related to dataset parameters and additional head - `mask_head`. I can use `MMDetectionArgsConverter` but with some fixes like:
```python
dataset = 'dataset.' if 'dataset' in cfg.data.train.keys() else ''
    update_config_map = {
        'train_ann_files': f'data.train.{dataset}ann_file',
        'train_data_roots': f'data.train.{dataset}img_prefix'}
```
```python
if 'mask_head' in cfg.model.roi_head.keys():
    update_config['model.roi_head.mask_head.num_classes'] = num_clusses
```
But I am not sure that this way is better.

File: pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py (hunk 7 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py:73]: Yes, from a configuration file it can be obtained but there must be additional updates: names and number of metric (`Bbox AP @ [IoU=0.50:0.95]`, `Segm AP @ [IoU=0.50:0.95]`) and a little bit different processing of resulting values. In my view via set of `if` statements it looks not so good.

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml (hunk 4 above)
Comment (evgeny-izutov) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml:2]: I would warn from creating omnipotent solutions... In this particular case the copy-paste is lesser evil - no potential troubles for others.

File: pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py (hunk 7 above)
Comment (evgeny-izutov) [reply pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py:73]: I would propose to parametrize this function with metrics (and `update_outputs`) and create two separate functions (aka `coco_ap_eval_det` and `coco_ap_eval_segm`) with call of generalized `coco_ap_eval` with appropriate metrics (and update rule).

Comment (evgeny-izutov) [reply pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py:73]: I would propose to parametrize this function with metrics (and `update_outputs`) and create two separate functions (aka `coco_ap_eval_det` and `coco_ap_eval_segm`) with call of generalized `coco_ap_eval` with appropriate metrics (and update rule).

File: pytorch_toolkit/instance_segmentation_2/tests/run_inference_tests.py (hunk 6 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/tests/run_inference_tests.py]: Removed.

File: pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+#    """ Collects average precision values in log file. """
+#
+#    average_precisions = []
+#    beginning = 'Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = '
+#    with open(path) as read_file:
+#        content = [line.strip() for line in read_file]
+#        for line in content:
+#            if line.startswith(beginning):
+#                average_precisions.append(float(line.replace(beginning, '')))
+#    return average_precisions
```
[the remark this answers is not in the document]
Comment (DmitriySidnev) [reply pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py]: Yes, forgot. Removed.

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml (hunk 4 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml:2]: @Ilya-Krylov, what do you think?

Comment (Ilya-Krylov) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml:2]: I don't think that it is omnipotent, I think that instance segmentation models and object detection models from mmdetection can be handled by one 'solution' though I do not insist.

Probably one more opinion, @druzhkov-paul ?

Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/modules.yaml]: Fixed. I applied `Repeat dataset` and added only `mask_head` section.

File: pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py (hunk 7 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/ote/ote/metrics/instance_segmentation/common.py]: Added two functions `coco_ap_eval_det` and `coco_ap_eval_segm` with common parametrized function `coco_ap_eval`. Part with independent metrics for instance segmentation removed.

File: pytorch_toolkit/ote/ote/modules/arg_converters/instance_segmentation.py (hunk 8 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/ote/ote/modules/arg_converters/instance_segmentation.py]: Done.

**Review by evgeny-izutov — APPROVED** [review]

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev please resolve conflicts

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml [hunk 10]
```diff
@@ -0,0 +1,62 @@
+name: instance-segmentation-0002
+domain: Instance Segmentation 2
+problem: COCO Instance Segmentation
+framework: OTEDetection
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml:4]: please set framework the same as in object detection models

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  destination: packages/ote
+- source: ../../requirements.txt
+  destination: requirements.txt
+max_nodes: 1
+training_target:
+- GPU
+inference_target:
+- CPU
+- iGPU
+- VPU
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml:30]: I guess exported model support neither iGPU nor VPU

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev unfortunately we are not able to run tests on CI, please do it locally from `pytorch_toolkit` folder.
<pre>
python tests/run_model_templates_tests.py
</pre>
They are failed

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml (hunk 10 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml]: Fixed.

File: pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml (hunk 11 above)
Comment (DmitriySidnev) [reply pytorch_toolkit/instance_segmentation_2/model_templates/coco-instance-segmentation/instance-segmentation-0002/template.yaml]: Remained only CPU.

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev please run tests, if not yet. The quantize.py script has been replaced by compress.py, change template file correspondingly please.

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, yes I already found out that `compess.py` replaced `quantize.py` but tests are failed on the face-detection with incorrect metrics. I'm checking it.

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, @LeonidBeynenson, for instance segmentation task are absent `nncf` tests but try to running that leads to failing. Should I add this test case like it is in `object detection`?

**Comment by Ilya-Krylov** [comment]
@DmitriySidnev As far as I know two-stage models are not supported yet by NNCF, you don't need to add NNCF tests, but please do whatever it is needed to make all tests pass

**Comment by DmitriySidnev** [comment]
@Ilya-Krylov, all bugs were fixed. All tests passed successfully.
